react/week3/src/index.js

```diff
@@ -0,0 +1,196 @@
+
+class Counter extends React.Component {
+  state = { counter: this.props.initialCounter };
+
+  componentDidMount() {
+    setInterval(() => {
+      this.setState({ counter: this.state.counter + 1 });
+    }, 1000);
+  }
+
+  render() {
+    return <div>
+       <p> You have used <strong> <i> {this.state.counter} </i> </strong> seconds on this website</p>
+   
+    </div>;
+  }
+}
+
+const TodoItems = (props) => {
+
+  const { items, deleteItem, todoCompleted, editItem } = props;
+  const length = items.length;
+  const showItems = length ? (
+    items.map((item) => {
+      return (
+        <div>
+          <li
+            key={item.id}
+            style={{ textDecoration: item.completed ? "line-through" : "none" }}
+          >
+            {item.description} | {item.deadline}
+            <input
+              type="checkbox"
+              checked={item.completed}
+              onChange={() => todoCompleted(item.id)}
+            />
+            <button onClick={() => deleteItem(item.id)}>Delete</button>
+            <button onClick={() => editItem(item.id)}>Edit</button>
+
+          </li>
+        </div>
+      );
+    })
+  ) : (
+      <p>NO items to show</p>
+    );
+  return <div>{showItems}</div>;
+};
+
+class AddItem extends React.Component {
+  state = {
+    description: "",
+    deadline: "",
+  };
+
+  handleChange = (e) => {
+    this.setState({
+      [e.target.id]: e.target.value,
+    });
+  };
+
+  handleSubmit = (e) => {
+    e.preventDefault();
+    console.log(this.state);
+    if (this.state.description === "" || this.state.deadline === "") {
+      console.log("false");
+    } else {
+      this.props.addItem(this.state);
+      this.setState({
+        description: "",
+        deadline: "",
+      });
+    }
+  };
+
+  render() {
+    return (
+      <div>
+        <form onSubmit={this.handleSubmit}>
+          <p>
+            Todo description :
+            <input  type="text"   placeholder="Enter description"   id="description"   onChange={this.handleChange}  value={this.state.description}  />
+          </p>
+          <p>
+            deadline :
+            <input type="date"  id="deadline"  onChange={this.handleChange}value={this.state.deadline}  />
+          </p>
+          <input className="submit" type="submit" value="Add Todo" />
+        </form>
+      </div>
+    );
+  }
+}
+
+function fetchTodo() {
+  const KEY = '391eee7a119b50bd2c5960ab51622532'
+  const URL = `https://gist.githubusercontent.com/benna100/${KEY}/raw`;
+  return fetch(URL).then((response) => response.json());
+}
+
+class App extends React.Component {
+  state = {
+    todos: [],
+  };
+
+  componentDidMount() {
+    fetchTodo().then((body) => {
+      this.setState({
+        todos: body,
+      });
+    });
+  }
+
+  deleteItem = (id) => {
+    console.log(id);
+    let items = this.state.todos.filter((item) => {
+      return item.id !== id;
+    });
+    this.setState({
+      todos: items,
+    });
```

Comment:
Same as above, watch how you modify the state - do all the necessary operations that rely on the previous state inside `setState`:

```
this.setState(prevState => {
    return {
        todos: prevState.filter(item => item.id !== id)
    }
})
```

Reply:
Ok